File: research/experiments/sweep/runs/opt-small-typescript-sonnet45/src/vec-ops.ts
/**
 * Pure vector arithmetic for n-dimensional optimization.
 * All operations return new arrays and never mutate inputs.
 */

/**
 * Dot product of two vectors
 */
export function dot(a: number[], b: number[]): number {
  let sum = 0;
  for (let i = 0; i < a.length; i++) {
    sum += a[i] * b[i];
  }
  return sum;
}

/**
 * Euclidean (L2) norm
 */
export function norm(v: number[]): number {
  return Math.sqrt(dot(v, v));
}

/**
 * Infinity norm (max absolute value)
 */
export function normInf(v: number[]): number {
  let max = 0;
  for (let i = 0; i < v.length; i++) {
    const abs = Math.abs(v[i]);
    if (abs > max) {
      max = abs;
    }
  }
  return max;
}

/**
 * Scalar multiplication
 */
export function scale(v: number[], s: number): number[] {
  const result = new Array(v.length);
  for (let i = 0; i < v.length; i++) {
    result[i] = v[i] * s;
  }
  return result;
}

/**
 * Element-wise addition
 */
export function add(a: number[], b: number[]): number[] {
  const result = new Array(a.length);
  for (let i = 0; i < a.length; i++) {
    result[i] = a[i] + b[i];
  }
  return result;
}

/**
 * Element-wise subtraction
 */
export function sub(a: number[], b: number[]): number[] {
  const result = new Array(a.length);
  for (let i = 0; i < a.length; i++) {
    result[i] = a[i] - b[i];
  }
  return result;
}

/**
 * Element-wise negation (scale(v, -1))
 */
export function negate(v: number[]): number[] {
  return scale(v, -1);
}

/**
 * Deep copy of a vector
 */
export function clone(v: number[]): number[] {
  return [...v];
}

/**
 * Create a vector of n zeros
 */
export function zeros(n: number): number[] {
  return new Array(n).fill(0);
}

/**
 * Fused operation: a + s*b (avoids intermediate allocation)
 */
export function addScaled(a: number[], b: number[], s: number): number[] {
  const result = new Array(a.length);
  for (let i = 0; i < a.length; i++) {
    result[i] = a[i] + s * b[i];
  }
  return result;
}

File: research/experiments/sweep/runs/opt-small-csharp-sonnet45/Optimization/VecOps.cs
namespace Optimization;

/// <summary>
/// Pure vector arithmetic for n-dimensional optimization.
/// All operations return new arrays and never mutate inputs.
/// </summary>
public static class VecOps
{
    /// <summary>Dot product of two vectors.</summary>
    public static double Dot(double[] a, double[] b)
    {
        double sum = 0;
        for (int i = 0; i < a.Length; i++)
        {
            sum += a[i] * b[i];
        }
        return sum;
    }

    /// <summary>Euclidean (L2) norm of a vector.</summary>
    public static double Norm(double[] v)
    {
        return Math.Sqrt(Dot(v, v));
    }

    /// <summary>Infinity norm (max absolute value) of a vector.</summary>
    public static double NormInf(double[] v)
    {
        double max = 0;
        foreach (double x in v)
        {
            double abs = Math.Abs(x);
            if (abs > max) max = abs;
        }
        return max;
    }

    /// <summary>Scalar multiplication.</summary>
    public static double[] Scale(double[] v, double s)
    {
        double[] result = new double[v.Length];
        for (int i = 0; i < v.Length; i++)
        {
            result[i] = v[i] * s;
        }
        return result;
    }

    /// <summary>Element-wise addition.</summary>
    public static double[] Add(double[] a, double[] b)
    {
        double[] result = new double[a.Length];
        for (int i = 0; i < a.Length; i++)
        {
            result[i] = a[i] + b[i];
        }
        return result;
    }

    /// <summary>Element-wise subtraction.</summary>
    public static double[] Sub(double[] a, double[] b)
    {
        double[] result = new double[a.Length];
        for (int i = 0; i < a.Length; i++)
        {
            result[i] = a[i] - b[i];
        }
        return result;
    }

    /// <summary>Element-wise negation.</summary>
    public static double[] Negate(double[] v)
    {
        return Scale(v, -1);
    }

    /// <summary>Deep copy of a vector.</summary>
    public static double[] Clone(double[] v)
    {
        return (double[])v.Clone();
    }

    /// <summary>Create a vector of n zeros.</summary>
    public static double[] Zeros(int n)
    {
        return new double[n];
    }

    /// <summary>Fused operation: a + s*b (avoids intermediate allocation).</summary>
    public static double[] AddScaled(double[] a, double[] b, double s)
    {
        double[] result = new double[a.Length];
        for (int i = 0; i < a.Length; i++)
        {
            result[i] = a[i] + s * b[i];
        }
        return result;
    }
}
